### PR TITLE
Adds REASSIGN OWNED BY propagation

### PR DESCRIPTION
**Pull request by gurkanindibay**
DESCRIPTION: Adds REASSIGN OWNED BY propagation

This pull request introduces the propagation of the "Reassign owned by" statement. It accommodates both local and distributed roles for both the old and new assignments. However, when the old role is a local role, it undergoes filtering and is not propagated. On the other hand, if the new role is a local role, the process involves first creating the role on worker nodes before propagating the "Reassign owned" statement.

### CODE REVIEW

**Comment by onurctirtir** [comment]
|                              |TO non-distributed role|TO distributed role|
|------------------------------|-----------------------|-------------------|
|FROM all-non-distributed roles|C1                     |C2                 |
|FROM some-distributed roles   |C3                     |C4                 |
|FROM all-distributed roles    |C5                     |C6                 |

Going through all the possible scenarios, it's obvious that while Citus should not return any DDLJobs for C1, it needs to return a DDLJob for C6.

However, I'm confused about the other scenarios, namely, C2, C3, C4 and C5 i.e., the cases where DDL both references non-distributed roles as well as distributed ones, either in `OWNED BY` clause or in `TO` clause.

As an example, for C4, would it really make sense to only reassign the objects owned by distributed ones to the one given in `TO` clause, or would it be confusing on user's end? We have two options here; either do what this PR proposes (filter the distributed roles in provided in `OWNED BY` claue and forget about the others), or automatically propagate non-distributed roles before sending `REASSIGN OWNED BY` command to workers.

I think choosing the first option would cause role objects to get out-of-sync on different nodes so might not be the ideal approach, so I'd like to talk more about this in weekly sync: @gurkanindibay, @JelteF.

Or, for C5, I'd expect the solution proposed in this PR to fail as it would incorrectly search for the role provided in `TO` clause on worker nodes. Do we have any tests for that?

File: src/backend/distributed/commands/owned.c [hunk 1]
```diff
@@ -88,3 +88,36 @@ PreprocessDropOwnedStmt(Node *node, const char *queryString,
...
+
+
+List *
+PreprocessReassignOwnedStmt(Node *node, const char *queryString,
+							ProcessUtilityContext processUtilityContext)
+{
+	ReassignOwnedStmt *stmt = castNode(ReassignOwnedStmt, node);
+	List *allReassignRoles = stmt->roles;
+
+	List *distributedReassignRoles = FilterDistributedRoles(allReassignRoles);
```
Comment (onurctirtir) [inline src/backend/distributed/commands/owned.c:100]: .

File: src/backend/distributed/commands/owned.c [hunk 2]
```diff
@@ -88,3 +88,36 @@ PreprocessDropOwnedStmt(Node *node, const char *queryString,
 
 	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
 }
+
+
+List *
+PreprocessReassignOwnedStmt(Node *node, const char *queryString,
```
Comment (onurctirtir) [inline src/backend/distributed/commands/owned.c:94]: Should we have this in Postprocess instead?

**Review by onurctirtir — CHANGES_REQUESTED** [review]
|                              |TO non-distributed role|TO distributed role|
|------------------------------|-----------------------|-------------------|
|FROM all-non-distributed roles|C1                     |C2                 |
|FROM some-distributed roles   |C3                     |C4                 |
|FROM all-distributed roles    |C5                     |C6                 |

Going through all the possible scenarios, it's obvious that while Citus should not return any DDLJobs for C1, it needs to return a DDLJob for C6.

However, I'm confused about the other scenarios, namely, C2, C3, C4 and C5 i.e., the cases where DDL both references non-distributed roles as well as distributed ones, either in `OWNED BY` clause or in `TO` clause.

As an example, for C4, would it really make sense to only reassign the objects owned by distributed ones to the one given in `TO` clause, or would it be confusing on user's end? We have two options here; either do what this PR proposes (filter the distributed roles in provided in `OWNED BY` claue and forget about the others), or automatically propagate non-distributed roles before sending `REASSIGN OWNED BY` command to workers.

I think choosing the first option would cause role objects to get out-of-sync on different nodes so might not be the ideal approach, so I'd like to talk more about this in weekly sync: @gurkanindibay, @JelteF.

Or, for C5, I'd expect the solution proposed in this PR to fail as it would incorrectly search for the role provided in `TO` clause on worker nodes. Do we have any tests for that?

**Comment by onurctirtir** [comment]
> REASSIGN	TO non-distributed role	TO distributed role
> OWNED BY all-non-distributed roles	C1	C2
> OWNED BY some-distributed roles	C3	C4
> OWNED BY all-distributed roles	C5	C6
> Going through all the possible scenarios, it's obvious that while Citus should not return any DDLJobs for C1, it needs to return a DDLJob for C6.
> 
> However, I'm confused about the other scenarios, namely, C2, C3, C4 and C5 i.e., the cases where DDL both references non-distributed roles as well as distributed ones, either in `OWNED BY` clause or in `TO` clause.
> 
> As an example, for C4, would it really make sense to only reassign the objects owned by distributed ones to the one given in `TO` clause, or would it be confusing on user's end? We have two options here; either do what this PR proposes (filter the distributed roles in provided in `OWNED BY` claue and forget about the others), or automatically propagate non-distributed roles before sending `REASSIGN OWNED BY` command to workers.
> 
> I think choosing the first option would cause role objects to get out-of-sync on different nodes so might not be the ideal approach, so I'd like to talk more about this in weekly sync: @gurkanindibay, @JelteF.
> 
> Or, for C5, I'd expect the solution proposed in this PR to fail as it would incorrectly search for the role provided in `TO` clause on worker nodes. Do we have any tests for that?

As we've discussed earlier this week, let's not filter out local roles but rely on workers to throw an error if the role doesn't exist on a worker node.

File: src/backend/distributed/commands/owned.c (hunk 2 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/owned.c:94]: There is no reason to move this to Postprocess actually. It just prepares the command. 
Normally when you check other statements who is in the nature of Alter like PreprocessAlterDatabaseStmt, we use Preprocess
Create needs to have Post since address only exists after local execution, Drop needs to be Pre since address is missing after local drop execution.
For Alter style statements, there is no such requirement.

File: src/backend/distributed/commands/owned.c (hunk 1 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/owned.c:100]: I could not understand what dot means?

Comment (onurctirtir) [reply src/backend/distributed/commands/owned.c:100]: Wanted to point out which portion of the code https://github.com/citusdata/citus/pull/7319#pullrequestreview-1715969930 is talking about, nvm :)

File: src/backend/distributed/commands/owned.c (hunk 2 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/owned.c:94]: let's get this into postprocess as we've discussed in group chat, due to reasons discussed there

Comment (onurctirtir) [reply src/backend/distributed/commands/owned.c:94]: > let's get this into postprocess as we've discussed in group chat, due to reasons discussed there

(discussion in group chat was still on-going indeed)

File: src/backend/distributed/commands/owned.c (hunk 1 above)
Comment (onurctirtir) [inline src/backend/distributed/commands/owned.c:100]: Should remove filtering logic as we've discussed two weeks ago.

File: src/backend/distributed/commands/owned.c (hunk 2 above)
Comment (onurctirtir) [inline src/backend/distributed/commands/owned.c:94]: need to move this down to Postprocess phase

File: src/test/regress/sql/reassure_owned.sql [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    pg_tables
+WHERE
+    tablename = 'test_table'
+  ) q2
+  $$
+) ORDER BY result;
+
+set citus.log_remote_commands to on;
+set citus.grep_remote_commands = '%REASSIGN OWNED BY%';
+REASSIGN OWNED BY role1 TO role2;
```
Comment (naisila) [inline src/test/regress/sql/reassure_owned.sql:28]: 1. Regardless of what we decide to do with the scenarios C1,C2,C3,C4,C5,C6, 
we should have at least one test for each scenario. And also we should ensure to have test cases with >1 role in the BY clause, e.g. 2 roles: `REASSIGN OWNED BY role0, role1 TO role2;`

2. As always, let's also add tests where the role names have escaped characters :)

**Review by naisila — COMMENTED** [review]
Since we haven't still decided yet on the logic, I only added some comments for the tests. Thank you.

File: src/backend/distributed/commands/owned.c (hunk 1 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/owned.c:120]: We decided to keep the filtering logic

File: src/backend/distributed/commands/owned.c (hunk 2 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/owned.c]: Fixed

File: src/backend/distributed/commands/owned.c (hunk 1 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/owned.c:120]: Added tests for distributed and non-distributed roles with escape characters

File: src/backend/distributed/commands/dependencies.c [hunk 4]
```diff
@@ -31,20 +31,65 @@
 #include "distributed/worker_manager.h"
 #include "distributed/worker_transaction.h"
 
+typedef enum RequiredObjectSet
+{
+	REQUIRE_ONLY_DEPENDENCIES = 1,
+	REQUIRE_OBJECT_AND_DEPENDENCIES = 2,
+} RequiredObjectSet;
+
```
Comment (naisila) [inline src/backend/distributed/commands/dependencies.c:39]: Naming can be quite tricky here :) First, we need a comment on this struct that explains clearly that this struct is created for function that ensures that objects exist on all nodes etc.
Here goes my naming suggestion:
```suggestion
typedef enum RequiredObjectSet
{
	ONLY_DEPENDENCIES = 1,
	DEPENDENCIES_AND_OBJECT = 2,
} ObjectsToExistOnAllNodes;

```

File: src/backend/distributed/commands/dependencies.c [hunk 5]
```diff
@@ -31,20 +31,65 @@
...
 
 static void EnsureDependenciesCanBeDistributed(const ObjectAddress *relationAddress);
 static void ErrorIfCircularDependencyExists(const ObjectAddress *objectAddress);
 static int ObjectAddressComparator(const void *a, const void *b);
 static void EnsureDependenciesExistOnAllNodes(const ObjectAddress *target);
+static void EnsureRequiredObjectExistOnAllNodes(const ObjectAddress *target,
```
Comment (naisila) [inline src/backend/distributed/commands/dependencies.c:45]: Naming suggestion: Let's use the name of the struct RequiredObjectSet inside the name of this function
```suggestion
static void EnsureRequiredObjectSetExistsOnAllNodes(const ObjectAddress *target,
```

File: src/backend/distributed/commands/dependencies.c [hunk 6]
```diff
@@ -55,29 +100,51 @@ static char * DropTableIfExistsCommand(Oid relationId);
...
  */
 static void
-EnsureDependenciesExistOnAllNodes(const ObjectAddress *target)
+EnsureRequiredObjectExistOnAllNodes(const ObjectAddress *target,
+									RequiredObjectSet requiredObjectSet)
 {
-	List *dependenciesWithCommands = NIL;
+	Assert(requiredObjectSet == REQUIRE_ONLY_DEPENDENCIES ||
+		   requiredObjectSet == REQUIRE_OBJECT_AND_DEPENDENCIES);
+
```
Comment (naisila) [inline src/backend/distributed/commands/dependencies.c:108]: Before doing anything else, here we can first check if the target object is in fact distributed: if it is  we can return early here, regardless of the `RequiredObjectSet`

File: src/backend/distributed/commands/owned.c [hunk 7]
```diff
@@ -89,3 +92,74 @@ PreprocessDropOwnedStmt(Node *node, const char *queryString,
...
+	stmt->roles = distributedReassignRoles;
+	char *sql = DeparseTreeNode((Node *) stmt);
+	stmt->roles = allReassignRoles;
+
+	ObjectAddress *newRoleAddress = GetNewRoleAddress(stmt);
+
+	int saveNestLevel = NewGUCNestLevel();
+	set_config_option("citus.enable_create_role_propagation", "on",
+					  (superuser() ? PGC_SUSET : PGC_USERSET), PGC_S_SESSION,
+					  GUC_ACTION_LOCAL, true, 0, false);
```
Comment (naisila) [inline src/backend/distributed/commands/owned.c:143]: We never do anything when calling EnsureDependencies function, and sometimes we may need to propagate roles in that function. Whatever we did before, the same can be applied here. So, I think we don't need these lines.

**Review by naisila — COMMENTED** [review]
Looking at the changes, I realize that it is much better to open a PR against this branch for the changed `EnsureDependencies` function. It was hard to follow in this PR. I already left comments here, but can you consider that? Especially if we will ask for an extra review from Nils or someone else.

File: src/backend/distributed/commands/owned.c (hunk 7 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/owned.c:143]: This block is required to set the GUC into previous state. I've used it in citus_internal_database_command
https://github.com/citusdata/citus/blob/18a8783e96983aef1c780d8019d5ce148b3dcad2/src/backend/distributed/metadata/metadata_sync.c#L4009

File: src/backend/distributed/commands/dependencies.c (hunk 4 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/dependencies.c:39]: Decided with @onurctirtir to keep as it is

File: src/backend/distributed/commands/dependencies.c [hunk 8]
```diff
@@ -55,29 +100,57 @@ static char * DropTableIfExistsCommand(Oid relationId);
  * postgres native CREATE IF NOT EXISTS, or citus helper functions.
  */
 static void
-EnsureDependenciesExistOnAllNodes(const ObjectAddress *target)
+EnsureRequiredObjectExistOnAllNodes(const ObjectAddress *target,
+									RequiredObjectSet requiredObjectSet)
 {
-	List *dependenciesWithCommands = NIL;
+	if (!IsAnyObjectDistributed(list_make1((ObjectAddress *) target)))
```
Comment (halilozanakgul) [inline src/backend/distributed/commands/dependencies.c:106]: We could use `IsObjectDistributed`

File: src/backend/distributed/commands/dependencies.c (hunk 6 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/dependencies.c:108]: I tested to put IsAnyObjectDistributed method however got test failures so I removed

File: src/backend/distributed/commands/dependencies.c (hunk 8 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/dependencies.c]: I needed to remove this line since it caused failure

File: src/backend/distributed/commands/owned.c [hunk 9]
```diff
@@ -89,3 +92,78 @@ PreprocessDropOwnedStmt(Node *node, const char *queryString,
...
+
+/*
+ * PostprocessReassignOwnedStmt Post-processes a REASSIGN OWNED statement.
+ *
+ * This function takes a Node pointer representing a REASSIGN OWNED statement,
+ * and performs any necessary post-processing after the statement has been executed.
+ * In this method, we propagate all the roles that are being reassigned.
+ * Therefore, we don't filter the roles based on whether they are distributed or not.
+ * If roles are not distributed and those roles are not present in the nodes
+ * where the statement is being propagated, then the statement will fail.
```
Comment (onurctirtir) [inline src/backend/distributed/commands/owned.c:105]:
```suggestion
 * PostprocessReassignOwnedStmt takes a Node pointer representing a REASSIGN
 * OWNED statement and performs any necessary post-processing after the statement
 * has been executed locally.
 *
 * We filter out local roles in OWNED BY clause before deparsing the command,
 * meaning that we skip reassigning what is owned by local roles. However,
 * if the role specified in TO clause is local, we automatically distribute
 * it before deparsing the command.
```

File: src/backend/distributed/commands/owned.c [hunk 10]
```diff
@@ -89,3 +92,78 @@ PreprocessDropOwnedStmt(Node *node, const char *queryString,
...
+
+	EnsureCoordinator();
+
+	stmt->roles = distributedReassignRoles;
+	char *sql = DeparseTreeNode((Node *) stmt);
+	stmt->roles = allReassignRoles;
+
+	ObjectAddress *newRoleAddress = GetNewRoleAddress(stmt);
+
+	int saveNestLevel = NewGUCNestLevel();
```
Comment (onurctirtir) [inline src/backend/distributed/commands/owned.c:133]:
```suggestion
	/*
	 * We temporarily enable create / alter role propagation to properly
	 * propagate the role specified in TO clause.
	 */
	int saveNestLevel = NewGUCNestLevel();
```

File: src/backend/distributed/commands/dependencies.c (hunk 6 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:108]: I believe lastly we agreed on adding the following check?

```suggestion
	if (IsAnyObjectDistributed(list_make1((ObjectAddress *) target)))
	{
		return;
	}
```

File: src/backend/distributed/commands/dependencies.c (hunk 5 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:45]: the suggestion makes sense to me

File: src/test/regress/sql/reassign_owned.sql [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+create ROLE local_target_role1;
+alter role local_target_role1 SET work_mem = '64MB';
+
+\c - - - :worker_1_port
+
+CREATE ROLE local_target_role1;
+
+\c - - - :master_port
+
+create role local_source_role1;
```
Comment (onurctirtir) [inline src/test/regress/sql/reassign_owned.sql:16]:
```suggestion
set citus.enable_create_role_propagation to off;
create role local_source_role1;
```

File: src/test/regress/sql/reassign_owned.sql [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+CREATE ROLE distributed_target_role1;
+
+set citus.enable_create_role_propagation to off;
+create ROLE local_target_role1;
+alter role local_target_role1 SET work_mem = '64MB';
+
+\c - - - :worker_1_port
+
+CREATE ROLE local_target_role1;
```
Comment (onurctirtir) [inline src/test/regress/sql/reassign_owned.sql:12]: Let's really make local_target_role1 a local role, right now it's distributed as far as I see in the test output.

```suggestion
set citus.enable_create_role_propagation to off;
CREATE ROLE local_target_role1;
```

File: src/test/regress/sql/reassign_owned.sql [hunk 13]
```diff
@@ -0,0 +1,140 @@
+CREATE ROLE distributed_source_role1;
+create ROLE "distributed_source_role-\!";
+
+CREATE ROLE distributed_target_role1;
+
+set citus.enable_create_role_propagation to off;
+create ROLE local_target_role1;
+alter role local_target_role1 SET work_mem = '64MB';
```
Comment (onurctirtir) [inline src/test/regress/sql/reassign_owned.sql:8]: Curiosity question: What is this testing btw?

File: src/test/regress/sql/reassign_owned.sql [hunk 14]
```diff
@@ -0,0 +1,140 @@
+CREATE ROLE distributed_source_role1;
+create ROLE "distributed_source_role-\!";
+
+CREATE ROLE distributed_target_role1;
```
Comment (onurctirtir) [inline src/test/regress/sql/reassign_owned.sql:4]: can we name this as such in order to test whether we properly quote target role?

```sql
"distributed_target_role1-\!""
```

File: src/test/regress/sql/reassign_owned.sql [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+WHERE
+    tablename in ('test_table', 'test_table2', 'test_table3', 'test_table4')
+  ) q2
+  $$
+) ORDER BY result;
+
+
+SET citus.log_remote_commands = true;
+set citus.grep_remote_commands = '%DROP ROLE%';
+drop role distributed_source_role1, "distributed_source_role-\!",distributed_target_role1,local_target_role1,local_source_role1;
```
Comment (onurctirtir) [inline src/test/regress/sql/reassign_owned.sql:140]: We don't need to be verbose when dropping the roles?

```suggestion
set client_min_messages to warning;
drop role distributed_source_role1, "distributed_source_role-\!",distributed_target_role1,local_target_role1,local_source_role1;
```

**Comment by onurctirtir** [comment]
Self-TODO: Test the PR manually once https://github.com/citusdata/citus/pull/7319#discussion_r1436055027 is resolved.

File: src/backend/distributed/commands/dependencies.c (hunk 6 above)
Comment (gurkanindibay) [reply src/backend/distributed/commands/dependencies.c:108]: This is the test error when we use IsAnyObjectDistributed(list_make1((ObjectAddress *) target))

File: src/test/regress/sql/reassign_owned.sql (hunk 14 above)
Comment (gurkanindibay) [reply src/test/regress/sql/reassign_owned.sql]: Applied suggestion

File: src/backend/distributed/deparser/deparse_owned_stmts.c [hunk 16]
```diff
@@ -83,3 +83,27 @@ AppendRoleList(StringInfo buf, List *roleList)
...
 }
+
+
+static void
+AppendReassignOwnedStmt(StringInfo buf, ReassignOwnedStmt *stmt)
+{
+	appendStringInfo(buf, "REASSIGN OWNED BY ");
+
+	AppendRoleList(buf, stmt->roles);
+	char const *newRoleName = RoleSpecString(stmt->newrole, true);
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/deparser/deparse_owned_stmts.c:94]: let's do `const char` instead of `char const` as we usually do then?

File: src/test/regress/expected/reassign_owned.out [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+set citus.log_remote_commands to on;
+set citus.grep_remote_commands = '%REASSIGN OWNED BY%';
+set citus.enable_create_role_propagation to off;
+set citus.enable_alter_role_propagation to off;
+set citus.enable_alter_role_set_propagation to off;
```
Comment (onurctirtir) [inline src/test/regress/expected/reassign_owned.out:117]: awesome test!

File: src/test/regress/sql/reassign_owned.sql [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+set citus.log_remote_commands to on;
+set citus.grep_remote_commands = '%REASSIGN OWNED BY%';
+set citus.enable_create_role_propagation to off;
+set citus.enable_alter_role_propagation to off;
+set citus.enable_alter_role_set_propagation to off;
+REASSIGN OWNED BY distributed_source_role1,"distributed_source_role-\!",local_source_role1 TO local_target_role1;
+
+show citus.enable_create_role_propagation;
+show citus.enable_alter_role_propagation;
+show citus.enable_alter_role_set_propagation;
```
Comment (onurctirtir) [inline src/test/regress/sql/reassign_owned.sql:94]: show stmts are not needed

File: src/backend/distributed/deparser/deparse_owned_stmts.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/deparser/deparse_owned_stmts.c:94]: please let's not forget changing this

**Review by onurctirtir — APPROVED** [review]
Looks very good to me, please consider those last two minor comments too, thanks!

File: src/test/regress/sql/reassign_owned.sql (hunk 18 above)
Comment (gurkanindibay) [reply src/test/regress/sql/reassign_owned.sql:94]: I'm rollbacking GUC state to previous one so I wanted to test GUC state if it is rollbacked

Comment (onurctirtir) [reply src/test/regress/sql/reassign_owned.sql:94]: oh that really makes sense
